### PR TITLE
Add rating service installation to the main installation flow

### PR DESCRIPTION
**Pull request by minherz**
Make the rating service provisioning to be integral part of the Hipster shop installation process.

The change removes conditional variable that prevented provisioning of the rating service when installing the Hipster shop using Terraform and/or Cloud Shell.
The rating service provisioning is converted into a terraform module in order to abstract the resource in the root module (03_ratingservice.tf).
The e2e test for the rating service is added to e2e latest workflow.
The frontend microservice environment is updated with RATING_SERVICE_ADDR variable which value is the "ratingservice" service of the App Engine application.

**NOTE:** I've added IAM roles to `test-runner@` SA so it will be able to provision additional resources that are required for the rating service.
**NOTE:** This change does not include modifications to frontend microservice to interact with the rating microservice.
**NOTE:** It is not possible to install the rating service using `skaffold run` command. Neither the frontend environment variable can be injected after the frontend is deployed. Using skaffold will require to have the rating service already be deployed. The skaffold does not support ([link1](https://github.com/GoogleContainerTools/skaffold/issues/2277), [link2](https://github.com/GoogleContainerTools/skaffold/issues/2960)) any custom actions on Deploy or after Deploy. As result, each call to `skaffold run` should be followed with the call to `kubectl set env` in order to inject RATING_SERVICE_ADDR environment variable.

### CODE REVIEW

**Review by simonz130 — APPROVED** [review]

File: .github/workflows/e2e-latest.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ jobs:
       run: |
         # clean GCP resources
         .github/workflows/e2e_scripts/clean.sh
-    - name: Build Sandbox Services
+    - name: Build Sandbox K8S Services
```
[the remark this answers is not in the document]
Comment (minherz) [reply .github/workflows/e2e-latest.yml:41]: OK

File: terraform/03_gke_cluster.tf [hunk 2]
```diff
@@ -141,7 +141,7 @@ data "google_compute_default_service_account" "default" {
 # Create GSA/KSA binding: let IAM auth KSAs as a svc.id.goog member name
 resource "google_service_account_iam_binding" "set_gsa_binding" {
   service_account_id = data.google_compute_default_service_account.default.name // google_service_account.set_gsa.name
-  role = "roles/iam.workloadIdentityUser"
+  role               = "roles/iam.workloadIdentityUser"
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/03_gke_cluster.tf:144]: Not my doing :( Probably automatic formatting in VSCode to align terraform variables.

File: terraform/03_ratingservice.tf [hunk 3]
```diff
@@ -12,116 +12,8 @@
...
-    # }
-
-    relative_uri = "/recollect"
-  }
-
-  depends_on = [google_project_service.cloudscheduler]
+module "ratingservice" {
+  source          = "./ratingservice"
+  gcp_project_id  = data.google_project.project.project_id
+  gcp_region_name = "us-east1"
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/03_ratingservice.tf:18]: 👍 Will do. loadgen is already implemented this way.

File: terraform/ratingservice/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+| variables.tf | Describes module input variables. |
+| output.tf | Describes rating service URL output variable. |
+
+The module provision additional Google APIs that in use by the rating service: `sql-component.googleapis.com` and `sqladmin.googleapis.com` to provision and configure the Postgres DB and `cloudscheduler.googleapis.com` to provision Cloud Scheduler task.
+
+### Bash scripts
+
+| Filename | Description |
+|---|---|
+| configure_rating_db.sh | Creates a DB schema in the Postgres DB and populates the DB with random generated data. In order to perform these operations from the local host, it temporarly adds the public IP of the host to the list of allowed IP of the CloudSQL instance. |
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/ratingservice/README.md:24]: When calling terraform destroy, the CloudSQL instance supposed to be deleted (see `deletion_protection = false` argument).

File: terraform/ratingservice/app.yaml.tpl [hunk 5]
```diff
@@ -0,0 +1,13 @@
+runtime: python38
+env: standard
+service: ${service_name}
+entrypoint: uwsgi --http-socket :8080 --wsgi-file main.py --callable app --master --processes 1 --threads ${max_connections}
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/ratingservice/app.yaml.tpl:4]: It is documented in /terraform/ratingservice/README.md file. Can you review and let me know if it is explanatory?

File: terraform/ratingservice/configure_rating_db.sh [hunk 6]
```diff
@@ -16,27 +16,32 @@
...
 # (machine has to have public ip)
 #
-MY_PUBLIC_IP=`dig TXT +short o-o.myaddr.l.google.com @ns1.google.com | awk -F'"' '{ print $2}'`
-gcloud sql instances patch $INSTANCE_NAME --authorized-networks=$MY_PUBLIC_IP --quiet
+echo "Launching Cloud SQL Proxy in background..."
+cloud_sql_proxy -instances=$DB_HOST=tcp:5432 -verbose=false &>/dev/null &
+cloud_proxy_pid=$!
+
+# wait until proxy establishes connection
+sleep 5
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/ratingservice/configure_rating_db.sh:35]: Yes, we can script a loop that tries to establish connection with psql and execute a command. I will refactor it in the following PRs (added as a tech debt).

File: terraform/ratingservice/main.tf [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+resource "google_project_service" "gae" {
+  service                    = "appengine.googleapis.com"
+  disable_dependent_services = true
+}
+
+resource "google_project_service" "cloudscheduler" {
+  service                    = "cloudscheduler.googleapis.com"
+  disable_dependent_services = true
+}
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/ratingservice/main.tf:42]: The description is in [TDD](https://docs.google.com/document/d/1k5Zq1BEyxhVe7Ko3_4dlLxhMXl4GC_tBp-6g1nOdnPg). Scheduler task periodically recollects the rating from new votes and updates the avg rating in the rating table.
I do not think that the terraform script is a right place to explain this. WDYT?

File: terraform/ratingservice/main.tf [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+resource "google_app_engine_application" "app" {
+  count       = data.external.app_engine_state.result.application_exist == "false" ? 1 : 0
+  location_id = var.gcp_region_name
+
+  depends_on = [google_project_service.gae]
+}
+
+resource "google_app_engine_standard_app_version" "default" {
```
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/ratingservice/main.tf:127]: It is impossible to deploy a non-default service -- a service with a custom name without having default service. I will add this explanation to README.md file.

File: terraform/ratingservice/main.tf (hunk 7 above)
[the remark this answers is not in the document]
Comment (minherz) [reply terraform/ratingservice/main.tf:42]: I have updated /terraform/ratingservice/README.md with more details about what resources are provisioned and why.

**Pull request merged**